<a href="https://colab.research.google.com/github/lakshitaison/firstproject/blob/main/breastcancer.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
import pandas as pd

df = pd.read_csv("breastcancer_updated.csv")
df.head()

,Age,Tumor Size,Regional Node Examined,Reginol Node Positive,Survival Months,Race_Other,Race_White,Marital Status_Married,Marital Status_Separated,Marital Status_Single,...,differentiate_Poorly differentiated,differentiate_Undifferentiated,differentiate_Well differentiated,Grade_2,Grade_3,Grade_anaplastic; Grade IV,A Stage_Regional,Estrogen Status_Positive,Progesterone Status_Positive,Node Ratio
0,43,40,19,11,1,True,False,True,False,False,...,False,False,False,True,False,False,True,True,True,0.578947
1,46,19,26,1,2,False,True,False,False,False,...,False,False,False,True,False,False,True,True,True,0.038462
2,67,25,4,1,2,False,True,True,False,False,...,True,False,False,False,True,False,True,True,True,0.250000
3,47,45,25,9,2,True,False,True,False,False,...,False,False,False,True,False,False,True,True,True,0.360000
4,55,15,9,1,3,False,True,True,False,False,...,False,False,False,True,False,False,True,True,True,0.111111


In [ ]:
X = df.drop("Status_Dead", axis=1)
y = df["Status_Dead"]

In [ ]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.2, random_state=42
)

In [ ]:
from sklearn.ensemble import RandomForestClassifier

model = RandomForestClassifier(n_estimators=100)

model.fit(X_train, y_train)

RandomForestClassifier()

In [ ]:
y_pred = model.predict(X_test)

In [ ]:
from sklearn.metrics import accuracy_score

accuracy = accuracy_score(y_test, y_pred)
print("Accuracy:", accuracy)

Accuracy: 0.9105590062111801


In [ ]:
output = pd.DataFrame({
    "Actual_Status": y_test,
    "Predicted_Status": y_pred
})

output.to_csv("results.csv", index=False)

In [ ]:
from google.colab import files
files.download("results.csv")

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

In [ ]:
output = X_test.copy()

output["Actual_Status"] = y_test.values
output["Predicted_Status"] = y_pred

output.to_csv("results.csv", index=False)

In [ ]:
from google.colab import files
files.download("results.csv")

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

**Reasoning**:
To save the cleaned DataFrame to a CSV file, I will use the `to_csv()` method, specifying the filename and `index=False` to prevent writing the DataFrame index.



In [ ]:
import pandas as pd

## Final Task

### Subtask:
Confirm that the processed dataset has been successfully saved as a CSV file.


# Task
Summarize the pandas code for creating a sample DataFrame, performing one-hot encoding on its categorical columns, and displaying the original and encoded DataFrames.

# Task
Summarize the pandas code used to load "breastcancer.xlsx", perform one-hot encoding on categorical columns, convert specified columns to numeric types, one-hot encode any remaining object columns, and save the resulting DataFrame to "breastcancer_processed.csv".

## Show Dataset Loading Code

### Subtask:
Display the pandas code used to load the 'breastcancer.xlsx' file into a DataFrame.


**Reasoning**:
To display the pandas code used for loading the 'breastcancer.xlsx' file into a DataFrame, I will use a code block to print the relevant Python code.



In [ ]:
print("""import pandas as pd\n\ndf = pd.read_excel('/content/breastcancer.xlsx')\ndf.head()""")

import pandas as pd

df = pd.read_excel('/content/breastcancer.xlsx')
df.head()


## Show Data Preprocessing Code

### Subtask:
Display the pandas code used for one-hot encoding specified categorical columns and converting numeric columns.


**Reasoning**:
To display the pandas code used for one-hot encoding specified categorical columns and converting numeric columns, I will use a code block to print the relevant Python code.



In [ ]:
print("""categorical_cols = ['Race', 'Marital Status', 'T Stage', 'N Stage', 'Status']
numeric_cols = ['Age', 'Tumor Size', 'Regional Node Examined', 'Reginol Node Positive']

# One-hot encode categorical columns
df = pd.get_dummies(df, columns=categorical_cols, drop_first=True)

# Convert specified columns to numeric, coercing errors
for col in numeric_cols:
    df[col] = pd.to_numeric(df[col], errors='coerce')

# Display the first 5 rows and data types to verify changes
print("DataFrame after preprocessing (first 5 rows):")
print(df.head())
print("\nDataFrame Info after preprocessing:")
df.info()""")

categorical_cols = ['Race', 'Marital Status', 'T Stage', 'N Stage', 'Status']
numeric_cols = ['Age', 'Tumor Size', 'Regional Node Examined', 'Reginol Node Positive']

# One-hot encode categorical columns
df = pd.get_dummies(df, columns=categorical_cols, drop_first=True)

# Convert specified columns to numeric, coercing errors
for col in numeric_cols:
    df[col] = pd.to_numeric(df[col], errors='coerce')

# Display the first 5 rows and data types to verify changes
print("DataFrame after preprocessing (first 5 rows):")
print(df.head())
print("
DataFrame Info after preprocessing:")
df.info()


# Task
Upload the "breastcancer.xlsx" file to the Colab environment.

## Upload Dataset

### Subtask:
Execute Python code to open a file selection dialog, allowing you to choose and upload your 'breastcancer.xlsx' file from your local machine to the Colab environment. The code will import `files` from `google.colab` and then call `files.upload()`.


**Reasoning**:
To allow the user to upload the 'breastcancer.xlsx' file, I will provide a Python code block that imports the `files` module from `google.colab` and then calls `files.upload()` to open a file selection dialog.



# Task
Develop a Random Forest Classifier to predict 'Status_Dead' using the "breastcancer.xlsx" dataset. This involves uploading the dataset, splitting it into features and target, applying a train-test split, training the classifier, evaluating its accuracy, and saving the model's predictions to a CSV file.

## Upload Dataset

### Subtask:
Upload the 'breastcancer.xlsx' file to the Colab environment.


**Reasoning**:
To allow the user to upload the 'breastcancer.xlsx' file, I will provide a Python code block that imports the `files` module from `google.colab` and then calls `files.upload()` to open a file selection dialog. This directly addresses the instructions provided for the subtask.



In [ ]:
from google.colab import files

uploaded = files.upload()

Saving breastcancer_updated.csv to breastcancer_updated.csv


## Split Dataset into X and y

### Subtask:
Separate the preprocessed DataFrame `df` into features `X` and target `y`, where 'Status_Dead' is identified as the target variable for prediction.


**Reasoning**:
To separate the preprocessed DataFrame into features `X` and target `y`, I will drop the 'Status_Dead' column from `df` to create `X`, and select the 'Status_Dead' column for `y`.



In [ ]:
X = df.drop('Status_Dead', axis=1)
y = df['Status_Dead']

print("Features (X) head:")
print(X.head())
print("\nTarget (y) head:")
print(y.head())

NameError: name 'df' is not defined

In [ ]:
# Load the uploaded breastcancer_updated.csv into a DataFrame
df = pd.read_csv('/content/breastcancer_updated.csv')
print('Dataset loaded successfully.')
display(df.head())

Dataset loaded successfully.


,Age,Tumor Size,Regional Node Examined,Reginol Node Positive,Survival Months,Race_Other,Race_White,Marital Status_Married,Marital Status_Separated,Marital Status_Single,...,differentiate_Poorly differentiated,differentiate_Undifferentiated,differentiate_Well differentiated,Grade_2,Grade_3,Grade_anaplastic; Grade IV,A Stage_Regional,Estrogen Status_Positive,Progesterone Status_Positive,Node Ratio
0,43,40,19,11,1,True,False,True,False,False,...,False,False,False,True,False,False,True,True,True,0.578947
1,46,19,26,1,2,False,True,False,False,False,...,False,False,False,True,False,False,True,True,True,0.038462
2,67,25,4,1,2,False,True,True,False,False,...,True,False,False,False,True,False,True,True,True,0.250000
3,47,45,25,9,2,True,False,True,False,False,...,False,False,False,True,False,False,True,True,True,0.360000
4,55,15,9,1,3,False,True,True,False,False,...,False,False,False,True,False,False,True,True,True,0.111111


In [ ]:
categorical_cols = ['Race', 'Marital Status', 'T Stage', 'N Stage', 'Status']
numeric_cols = ['Age', 'Tumor Size', 'Regional Node Examined', 'Reginol Node Positive']

# One-hot encode categorical columns
df = pd.get_dummies(df, columns=categorical_cols, drop_first=True)

# Convert specified columns to numeric, coercing errors
for col in numeric_cols:
    df[col] = pd.to_numeric(df[col], errors='coerce')

# Drop any remaining object columns that weren't specified in categorical_cols
# This is a robust way to handle any other unexpected categorical columns after initial processing
for col in df.select_dtypes(include=['object']).columns:
    df = pd.get_dummies(df, columns=[col], drop_first=True)

print('Preprocessing complete. First 5 rows of the processed DataFrame:')
display(df.head())
print('\nDataFrame Info after preprocessing:')
df.info()

KeyError: "None of [Index(['Race', 'Marital Status', 'T Stage', 'N Stage', 'Status'], dtype='object')] are in the [columns]"

In [ ]:
# Split the preprocessed DataFrame into features (X) and target (y)
X = df.drop('Status_Dead', axis=1)
y = df['Status_Dead']

print('Features (X) head:')
display(X.head())
print('\nTarget (y) head:')
display(y.head())
print(f"X shape: {X.shape}")
print(f"y shape: {y.shape}")

Features (X) head:


,Age,Tumor Size,Regional Node Examined,Reginol Node Positive,Survival Months,Race_Other,Race_White,Marital Status_Married,Marital Status_Separated,Marital Status_Single,...,differentiate_Poorly differentiated,differentiate_Undifferentiated,differentiate_Well differentiated,Grade_2,Grade_3,Grade_anaplastic; Grade IV,A Stage_Regional,Estrogen Status_Positive,Progesterone Status_Positive,Node Ratio
0,43,40,19,11,1,True,False,True,False,False,...,False,False,False,True,False,False,True,True,True,0.578947
1,46,19,26,1,2,False,True,False,False,False,...,False,False,False,True,False,False,True,True,True,0.038462
2,67,25,4,1,2,False,True,True,False,False,...,True,False,False,False,True,False,True,True,True,0.250000
3,47,45,25,9,2,True,False,True,False,False,...,False,False,False,True,False,False,True,True,True,0.360000
4,55,15,9,1,3,False,True,True,False,False,...,False,False,False,True,False,False,True,True,True,0.111111



Target (y) head:


,Status_Dead
0,False
1,True
2,True
3,False
4,False


X shape: (4024, 30)
y shape: (4024,)


## Apply Train-Test Split

### Subtask:
Divide the feature set (`X`) and target variable (`y`) into training and testing subsets (`X_train`, `X_test`, `y_train`, `y_test`) to prepare the data for model training and evaluation.


**Reasoning**:
To apply the train-test split, I need to import the `train_test_split` function from `sklearn.model_selection`, then use it to split `X` and `y` with the specified `test_size` and `random_state`, and finally print the shapes of the resulting sets for verification.



In [ ]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

print(f"X_train shape: {X_train.shape}")
print(f"X_test shape: {X_test.shape}")
print(f"y_train shape: {y_train.shape}")
print(f"y_test shape: {y_test.shape}")

X_train shape: (3219, 30)
X_test shape: (805, 30)
y_train shape: (3219,)
y_test shape: (805,)


## Train Random Forest Classifier

### Subtask:
Initialize and train a Random Forest Classifier model using the training data (`X_train` and `y_train`).


**Reasoning**:
To initialize and train a Random Forest Classifier, I need to import the necessary class, instantiate it, and then fit the model to the training data (`X_train` and `y_train`).



In [ ]:
from sklearn.ensemble import RandomForestClassifier

# Initialize the Random Forest Classifier
model = RandomForestClassifier(random_state=42)

# Train the model
model.fit(X_train, y_train)

print("Random Forest Classifier trained successfully.")

Random Forest Classifier trained successfully.


## Train Random Forest Classifier

### Subtask:
Initialize and train a Random Forest Classifier model using the training data (`X_train` and `y_train`).

#### Instructions
1. Import the `RandomForestClassifier` class from `sklearn.ensemble`.
2. Instantiate a `RandomForestClassifier` object. You can use default parameters or specify `random_state` for reproducibility (e.g., `random_state=42`).
3. Fit the instantiated Random Forest Classifier model to the training data using the `.fit()` method, passing `X_train` and `y_train`.

## Evaluate Model Accuracy

### Subtask:
Make predictions on the test set (`X_test`) using the trained Random Forest Classifier and evaluate the model's accuracy.

#### Instructions
1. Use the trained `model` to make predictions on `X_test` and store them in a variable like `y_pred`.
2. Import the `accuracy_score` metric from `sklearn.metrics`.
3. Calculate the accuracy of the model by comparing `y_pred` with `y_test`.
4. Print the calculated accuracy.

**Reasoning**:
Now that the Random Forest Classifier has been trained, I will make predictions on the test set and evaluate the model's accuracy as per the instructions in the subtask.



In [ ]:
from sklearn.metrics import accuracy_score

# Make predictions on the test set
y_pred = model.predict(X_test)

# Calculate the accuracy
accuracy = accuracy_score(y_test, y_pred)

print(f"Model Accuracy: {accuracy:.4f}")

Model Accuracy: 0.9130


## Save Predictions to CSV

### Subtask:
Save the model's predictions (`y_pred`) for the test set to a new CSV file named `predictions.csv`. This file should contain a single column for the predictions.

#### Instructions
1. Create a pandas DataFrame from the `y_pred` array. Name the column appropriately, e.g., 'Predicted_Status_Dead'.
2. Define the `output_filepath` for the predictions CSV file (e.g., `'/content/predictions.csv'`).
3. Use the `to_csv()` method of the DataFrame to save it to the specified `output_filepath`, ensuring `index=False` to prevent writing the DataFrame index to the CSV.
4. Print a confirmation message indicating that the predictions have been saved and providing the file path.

**Reasoning**:
To save the model's predictions to a CSV file, I will create a pandas DataFrame from the `y_pred` array, define an output filepath, and then use the `to_csv()` method with `index=False`.



In [ ]:
import pandas as pd

# Create a DataFrame from the predictions
predictions_df = pd.DataFrame(y_pred, columns=['Predicted_Status_Dead'])

# Define the output filepath
predictions_output_filepath = '/content/predictions.csv'

# Save the predictions to a CSV file
predictions_df.to_csv(predictions_output_filepath, index=False)

print(f"Predictions saved to {predictions_output_filepath}")

Predictions saved to /content/predictions.csv


## Final Task

### Subtask:
Summarize the model's performance, including the calculated accuracy, and confirm that the predictions have been successfully saved to a CSV file for further use.


## Summary:

### Q&A
*   The model's performance, specifically its accuracy, is 0.9130.
*   The predictions have been successfully saved to a CSV file named `predictions.csv` at the path `/content/predictions.csv`.

### Data Analysis Key Findings
*   The dataset, although initially specified as `breastcancer.xlsx`, was uploaded as `breastcancer_updated (4).csv`.
*   The data was successfully split into features (`X`) and target (`y`), with 'Status_Dead' as the target variable.
*   The dataset was divided into training (3219 samples) and testing (805 samples) sets using an 80/20 split and `random_state=42`.
*   A Random Forest Classifier was initialized and trained successfully.
*   The trained Random Forest Classifier achieved an accuracy of 0.9130 on the test set.
*   The model's predictions for the test set were saved to a CSV file named `predictions.csv`.

### Insights or Next Steps
*   Given the high accuracy of 0.9130, the Random Forest model shows strong performance in predicting 'Status\_Dead'.
*   Further analysis could include evaluating other metrics (e.g., precision, recall, F1-score) and exploring hyperparameter tuning to potentially enhance model performance.
